### PR TITLE
Do not overwrite `"path"` and `"root_path"` scope keys

### LINKED ISSUE #1336: Handling of `scope["root_path"]` and `scope["path"]` differs from ASGI spec
### Checklist

- [x] The bug is reproducible against the latest release and/or `master`.
- [x] There are no similar issues or pull requests to fix it yet.

### Describe the bug

Accoring to https://github.com/django/asgiref/issues/229#issuecomment-765583185, `scope["path"]` includes the `scope["root_path"]`.
So if root_path is `/foo` and request for `/foo/bar` has come, the upstream application servers (such as uvicorn) are passing `scope["path"] == "/foo/bar" and scope["root_path"] == "/foo"`, which is correct behavior. 

But starlette does not conform this. It assumes `scope["path"]` is a remainder of stripping `scope["root_path"]` from original request path. In this case, `scope["path"] == "/bar" and scope["root_path"] == "/foo"`. This ASGI incompatible assumption can be seen [here](https://github.com/encode/starlette/blob/6af5c515e0a896cbf3f86ee043b88f6c24200bcf/starlette/routing.py#L368) or [here](https://github.com/encode/starlette/blob/6af5c515e0a896cbf3f86ee043b88f6c24200bcf/starlette/datastructures.py#L23) and many other places.

The following "To reproduce" section shows just one aspect of this wrong assumption.

### To reproduce

1. Save `main.py` as follows.
```python
from starlette.applications import Starlette
from starlette.responses import PlainTextResponse
from starlette.routing import Route, Mount


class Bar:
    async def __call__(self, scope, receive, send):
        await PlainTextResponse(f"bar {scope['root_path']=} {scope['path']=}")(scope, receive, send)

class FooBar:
    async def __call__(self, scope, receive, send):
        await PlainTextResponse(f"foobar {scope['root_path']=} {scope['path']=}")(scope, receive, send)

routes =[
    Route('/bar', endpoint=Bar()),
    Mount('/foo', routes=[
        Route('/bar', endpoint=FooBar())
    ])
]
app = Starlette(routes=routes)
```

2. Run `uvicorn main:app --port 8000 --root-path /foo`
3. Access by `curl http://localhost:8000/foo/bar`

### Expected behavior

Receives
```
bar scope['root_path']='/foo' scope['path']='/foo/bar'
```

### Actual behavior

Receives
```
foobar scope['root_path']='/foo/foo' scope['path']='/bar'
```

Some points here are,
1. `scope['path']` does not include `scope['root_path']` , which is ASGI incompatible.
2. `FooBar` ASGI handler is called. It does not take `scope['root_path']` into account when routing.

### Environment

- OS: Linux
- Python version: 3.8.2
- Starlette version: 0.17.0

### Additional context

This may be a root cause of [this issue in fastapi](https://github.com/tiangolo/fastapi/issues/2872)

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/starlette/issues/1336">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/starlette/issues/1336/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by Kludex**
- Closes #1336
- Completes #1774

### CODE REVIEW

**Comment by Kludex** [comment]
I'm going to try to make this PR cleaner, but it already solves the problem.

File: starlette/routing.py [hunk 1]
```diff
@@ -402,21 +410,23 @@ def routes(self) -> typing.List[BaseRoute]:
...
-                matched_path = path[: -len(remaining_path)]
+                matched_path = current_path[: -len(remaining_path)]
                 path_params = dict(scope.get("path_params", {}))
                 path_params.update(matched_params)
                 root_path = scope.get("root_path", "")
                 child_scope = {
                     "path_params": path_params,
                     "app_root_path": scope.get("app_root_path", root_path),
-                    "root_path": root_path + matched_path,
-                    "path": remaining_path,
```
Comment (frankie567) [inline starlette/routing.py:429]: Don't we need to keep the original `root_path` and `path` in the child scope?

**Review by frankie567 — COMMENTED** [review]
Apart from my question above, LGTM! You have filled the holes I left in my previous PR :) 

Still not convinced about the naming `current_root_path` and `current_path`, but that's minor 😄 Suggestion (not ideal): `route_root_path` / `route_path`.

File: starlette/routing.py (hunk 1 above)
Comment (Kludex) [reply starlette/routing.py:429]: Aren't they always there anyway? Those lines were just modifying them, no?

Comment (frankie567) [reply starlette/routing.py:429]: Yes, you're right. My point was if we were using this `child_scope` as the new scope for inner routers, but it seems actually that it's always merged with the root `scope`.

**Review by frankie567 — APPROVED** [review]

**Comment by Kludex** [comment]
@encode/maintainers can someone review this, please?

**Review by abersheeran — APPROVED** [review]
It looks very nice, I like the regular expressions here.

**Pull request merged**